jsonrpc2/codecs_test.go

```diff
@@ -2,39 +2,127 @@ package jsonrpc2
 
 import (
 	"bytes"
+	"encoding/json"
 	"io"
-	"io/ioutil"
-	"reflect"
+	"sort"
 	"testing"
 )
 
+type nopCloser struct {
+	io.ReadWriter
+}
+
+func (nopCloser) Close() error { return nil }
+
 func TestCodec(t *testing.T) {
 	var buf bytes.Buffer
-	rwc := struct {
-		io.Reader
-		io.Writer
-		io.Closer
-	}{
-		Reader: &buf,
-		Writer: &buf,
-		Closer: ioutil.NopCloser(&buf),
-	}
-
-	codec := IOCodec(rwc)
+	codec := IOCodec(nopCloser{&buf})
 	msg := &Message{
-		ID:      []byte("42"),
-		Version: "2.0",
+		ID: []byte("42"),
+		Request: &Request{
+			Method: "foo",
+		},
 	}
 	err := codec.WriteMessage(msg)
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	got, want := buf.String(), `{"method":"foo","id":42}`+"\n"
+	if got != want {
+		t.Errorf("got: %q; want %q", got, want)
+	}
+
 	msg2, err := codec.ReadMessage()
 	if err != nil {
 		t.Error(err)
 	}
 
-	if !reflect.DeepEqual(msg, msg2) {
-		t.Errorf("got: %q; want %q", msg2, msg)
+	assertEqualJSON(t, msg, msg2, "write/read compare")
+}
+
+func TestBatch(t *testing.T) {
+	tests := []struct {
+		In    string
+		Want  []Message
+		Error bool
+	}{
+		{
+			In: `[{"method": "foo", "id": 1}]`,
+			Want: []Message{
+				{ID: json.RawMessage("1"), Request: &Request{Method: "foo"}},
+			},
+		},
+		{
+			In: `
+				[
+					{"method": "foo", "id": 2},
+					{"method": "bar", "id": 3}
+				]
+			`,
+			Want: []Message{
+				{ID: json.RawMessage("2"), Request: &Request{Method: "foo"}},
+				{ID: json.RawMessage("3"), Request: &Request{Method: "bar"}},
+			},
+		},
+		{
+			In:    `[]`,
+			Error: true,
+		},
+		{
+			In: `
+				[
```

Comment:
Is this actually desired behavior?

Reply:
@ryanschneider How do you mean?

I'm probably missing something, but looks sensible to me. Note that the input is a JSON stream, so multiple messages are supported.